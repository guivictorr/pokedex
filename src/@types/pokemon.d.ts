import { StatNames } from '../components/Statistic';

export type StatKeys = keyof typeof StatNames;

type Types = {
  type: {
    name: string;
    url: string;
  };
};

type Sprites = {
  front_default: string;
  back_default: string;
};

type Stats = {
  base_stat: number;
  effort: number;
  stat: {
    name: StatKeys;
    url: string;
  };
};

type PokemonProps = {
  id: number;
  name: string;
  sprites: Sprites;
  types: Types[];
  stats: Stats[];
  weight: number;
  height: number;
};

export type PokemonsByLimit = {
  count: number;
  next: string | null;
  previous: string | null;
  results: {
    name: string;
    url: string;
  }[];
};

export default PokemonProps;
