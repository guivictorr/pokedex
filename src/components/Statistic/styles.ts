import styled from 'styled-components';

type FillerProps = {
  percentage: number;
};

export const Container = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  max-width: 310px;

  p {
    font-size: ${({ theme }) => theme.fontSizes.xsmall};
  }
`;

export const ProgressBar = styled.div`
  position: relative;
  height: 10px;
  width: 100%;
  max-width: 250px;
  border-radius: 8px;
  margin: 0 auto;
  background-color: ${({ theme }) => theme.colors.grey200};
`;

export const Filler = styled.div<FillerProps>`
  background-color: ${({ theme }) => theme.colors.primary};
  height: 100%;
  width: ${({ percentage }) => percentage}%;
  border-radius: inherit;
`;
