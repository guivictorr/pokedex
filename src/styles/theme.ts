export default {
  colors: {
    primary: 'red',
  },
};
