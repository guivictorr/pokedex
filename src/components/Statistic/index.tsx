import * as S from './styles';

type StatisticProps = {
  percentage: number;
  label: StatKeys;
};

export enum StatNames {
  hp = 'HP',
  attack = 'ATK',
  defense = 'DEF',
  'special-attack' = 'S.ATK',
  'special-defense' = 'S.DEF',
  speed = 'SPD',
}

export type StatKeys = keyof typeof StatNames;

const Statistic = ({ percentage, label }: StatisticProps) => {
  return (
    <S.Container>
      <p>{StatNames[label]}</p>
      <S.ProgressBar>
        <S.Filler
          data-testid="filler"
          initial={{ width: 0 }}
          animate={{ width: percentage > 100 ? '100%' : percentage + '%' }}
          transition={{ duration: 1 }}
          percentage={percentage}
        />
      </S.ProgressBar>
      <p>{percentage}</p>
    </S.Container>
  );
};

export default Statistic;
