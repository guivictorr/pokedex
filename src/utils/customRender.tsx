import { ReactNode } from 'react';
import { render } from '@testing-library/react';
import { AnimateSharedLayout } from 'framer-motion';
import { ThemeProvider } from 'styled-components';
import GlobalStyles from '../styles/global';
import FavoritesProvider from '../context/favoriteContext';
import ModalProvider from '../context/modalContext';
import theme from '../styles/theme';

const customRender = (children: ReactNode) => {
  return render(
    <ThemeProvider theme={theme}>
      <GlobalStyles />

      <ModalProvider>
        <FavoritesProvider>
          <AnimateSharedLayout type="crossfade">{children}</AnimateSharedLayout>
        </FavoritesProvider>
      </ModalProvider>
    </ThemeProvider>,
  );
};

export default customRender;
