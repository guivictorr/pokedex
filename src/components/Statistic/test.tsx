import { screen } from '@testing-library/react';
import Statistic, { StatNames } from '.';
import { StatKeys } from '../../@types/pokemon';
import theme from '../../styles/theme';
import customRender from '../../utils/customRender';

describe('<Statistic />', () => {
  it('should render correctly', async () => {
    const { container } = customRender(
      <Statistic percentage={100} label="defense" />,
    );

    expect(container.firstChild).toMatchSnapshot();
    const filler = screen.getByTestId('filler');

    expect(filler).toHaveStyle({
      'background-color': theme.colors.primary,
    });
  });

  it('should render correct background color when pass 100', () => {
    customRender(<Statistic percentage={150} label="defense" />);

    const filler = screen.getByTestId('filler');

    expect(filler).toHaveStyle({
      'background-color': theme.colors.danger,
    });
  });

  it('should render statistic with correct labels and values', () => {
    const mocks: { label: StatKeys; percentage: number }[] = [
      { label: 'hp', percentage: 100 },
      { label: 'attack', percentage: 24 },
      { label: 'defense', percentage: 54 },
      { label: 'special-attack', percentage: 21 },
      { label: 'special-defense', percentage: 32 },
      { label: 'speed', percentage: 6 },
    ];

    customRender(
      mocks.map(({ label, percentage }, index) => (
        <Statistic key={`test${index}`} percentage={percentage} label={label} />
      )),
    );

    mocks.forEach(({ percentage, label }) => {
      expect(screen.getByText(percentage)).toBeInTheDocument();
      expect(screen.getByText(StatNames[label])).toBeInTheDocument();
    });
  });
});
