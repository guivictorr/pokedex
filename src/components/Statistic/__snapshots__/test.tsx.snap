// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Statistic /> should render correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  max-width: 350px;
  margin: 8px 0;
}

.c0 p {
  font-size: 12px;
  width: 32px;
}

.c1 {
  height: 10px;
  width: 100%;
  max-width: 250px;
  border-radius: 8px;
  background-color: #d2d4d6;
}

.c2 {
  background-color: #FFCB05;
  height: 100%;
  border-radius: inherit;
}

<div
  class="c0"
>
  <p>
    DEF
  </p>
  <div
    class="c1"
  >
    <div
      class="c2"
      data-testid="filler"
      style="width: 0px;"
    />
  </div>
  <p>
    100
  </p>
</div>
`;
